q2galaxy/core/usage.py

```diff
@@ -105,15 +108,50 @@ def to_interface_name(self, skip_ref=False):
             return '.'.join([self.prefix, name])
         return name
 
-    def assert_output_type(self, semantic_type):
+    def assert_output_type(self, semantic_type, key=None):
         semantic_type = re.escape(str(semantic_type))
-        self._galaxy_has_line_matching(path='metadata.yaml',
+        path = 'metadata.yaml'
+
+        if key is not None:
+            key = str(key)
+
+        if self.var_type == 'result_collection' and key:
+            self._key_helper(path, expression='type: EchoOutput', key=key)
+            return
+
+        self._galaxy_has_line_matching(path=path,
                                        expression=f'type: {semantic_type}')
 
-    def assert_has_line_matching(self, path, expression):
-        self._galaxy_has_line_matching(path=f'data\\/{path}',
+    def assert_has_line_matching(self, path, expression, key=None):
+        path = f'data\\/{path}'
+
+        if key is not None:
+            key = str(key)
+
+        if self.var_type == 'result_collection' and key:
+            self._key_helper(path, expression, key)
+            return
+
+        self._galaxy_has_line_matching(path=path,
                                        expression=expression)
 
+    def _key_helper(self, path, expression, key):
+        output = self.use.output_lookup[self.name]
+        element = XMLNode('element', name=str(key), ftype="qza")
```

Comment:
will `str(key)` actually be the combined path like `/output_dir/something.qza`? It looks like it would just be `something.qza`. If that's what the driver emits, then there's potential for overwriting other outputs as they aren't namespaced by the collection.

Reply:
Looks like the driver does just put it in a directory, so I think this element node needs a combined file-path for the name.

Reply:
That codepath was never being taken because it was still looking for the old 'ResultCollection' type, so the tests weren't quite templating the way I intended, but I hadn't noticed until now because they were still passing. I changed it so the tests are now templating in the desired manner, and the tests are passing with things as they are in that _key_helper functions. I believe the name in that element tag is specifically the name of the element within the output collection. The directory its saved to is basically implied.

The test assertions look like this:

```
<test>
    <output_collection name="output" type="list">
        <assert_contents>
            <has_archive_member path="[0-9a-f]{8}-[0-9a-f]{4}-[4][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\/metadata.yaml">
                <has_line_matching expression="type: Collection\[EchoOutput\]"/>
            </has_archive_member>
        </assert_contents>
        <element name="42" ftype="qza">
            <assert_contents>
                <has_archive_member path="[0-9a-f]{8}-[0-9a-f]{4}-[4][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\/metadata.yaml">
                    <has_line_matching expression="type: EchoOutput"/>
                </has_archive_member>
            </assert_contents>
        </element>
        <element name="42" ftype="qza">
            <assert_contents>
                <has_archive_member path="[0-9a-f]{8}-[0-9a-f]{4}-[4][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\/data\/echo.txt">
                    <has_line_matching expression="0: 42"/>
                </has_archive_member>
            </assert_contents>
        </element>
        <element name="43" ftype="qza">
            <assert_contents>
                <has_archive_member path="[0-9a-f]{8}-[0-9a-f]{4}-[4][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\/metadata.yaml">
                    <has_line_matching expression="type: EchoOutput"/>
                </has_archive_member>
            </assert_contents>
        </element>
        <element name="43" ftype="qza">
            <assert_contents>
                <has_archive_member path="[0-9a-f]{8}-[0-9a-f]{4}-[4][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\/data\/echo.txt">
                    <has_line_matching expression="0: 43"/>
                </has_archive_member>
            </assert_contents>
        </element>
    </output_collection>
</test>
```

Reply:
Now looks like this. Key takeaways are that the first `assert_contents` outside an `element` tag in the prior example doesn't do anything and appears to be ignored and having multiple `element` tags with the same name or even multiple `assert_contents` tags within an `element` does not produce the desired behavior. They need to be properly nested not in serial.

```
        <test>
            <output_collection name="output" type="list">
                <element name="42" ftype="qza">
                    <assert_contents>
                        <has_archive_member path="[0-9a-f]{8}-[0-9a-f]{4}-[4][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\/metadata.yaml">
                            <has_line_matching expression="type: EchoOutput"/>
                        </has_archive_member>
                        <has_archive_member path="[0-9a-f]{8}-[0-9a-f]{4}-[4][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\/data\/echo.txt">
                            <has_line_matching expression="0: 42"/>
                        </has_archive_member>
                    </assert_contents>
                </element>
                <element name="43" ftype="qza">
                    <assert_contents>
                        <has_archive_member path="[0-9a-f]{8}-[0-9a-f]{4}-[4][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\/metadata.yaml">
                            <has_line_matching expression="type: EchoOutput"/>
                        </has_archive_member>
                        <has_archive_member path="[0-9a-f]{8}-[0-9a-f]{4}-[4][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\/data\/echo.txt">
                            <has_line_matching expression="0: 43"/>
                        </has_archive_member>
                    </assert_contents>
                </element>
            </output_collection>
        </test>
```